packages/flutter_tools/lib/src/isolated/resident_web_runner.dart

```diff
@@ -303,6 +303,27 @@ Please provide a valid TCP port (an integer between 0 and 65535, inclusive).
                 ? WebExpressionCompiler(device!.generator!, fileSystem: _fileSystem)
                 : null;
 
+        // Retrieve connected web devices, excluding the web server device.
+        final List<Device>? devices = await globals.deviceManager?.getAllDevices();
+        final Set<String> supportedWebDeviceIds =
+            devices
+                ?.where(
+                  (Device d) =>
+                      d.platformType == PlatformType.web &&
+                      d.isConnected &&
+                      d.id != WebServerDevice.kWebServerDeviceId,
+                )
+                .map((Device d) => d.id)
+                .toSet() ??
+            <String>{};
+
+        // If no web devices are connected, or the specified device does not support full debugging (ie. web-server),
+        // we should use DWDS WebSocket connection instead of Chrome-based connection. If no device is specified,
+        // we fall back to the default Chrome connection.
+        final bool useDwdsWebSocketConnection =
+            supportedWebDeviceIds.isEmpty ||
+            !supportedWebDeviceIds.contains(globals.deviceManager?.specifiedDeviceId ?? 'chrome');
```

Comment:
nit: `GoogleChromeDevice.kChromeDeviceId` instead of `chrome`

Reply:
This logic falling back on chrome here is still slightly confusing to me. Is this effectively the same as:

```
final bool useDwdsWebSocketConnection =
     supportedWebDeviceIds.isEmpty ||
     (globals.deviceManager?.specifiedDeviceId != null &&
     !supportedWebDeviceIds.contains(globals.deviceManager!.specifiedDeviceId!));
```